spec/registry.md

```diff
@@ -0,0 +1,162 @@
+# WIP DRAFT MessageFormat 2.0 Registry
+
+_This document is non-normative._
+
+The implementations and tooling can greatly benefit from a structured definition of formatting and matching functions available to messages at runtime.
+The _registry_ is a mechanism for storing such declarations in a portable manner.
+
+## Goals
+
+The goal of the registry is to provide a machine-readable description of MessageFormat extensions (custom functions) in order to:
+
+* support the localization roundtrip,
+* validate message correctness, and
+* improve the authoring experience.
+
+## Use Cases
+
+The registry enables the following usage scenarios:
+
+* Generate variant keys for a given locale during XLIFF extraction.
+* Verify the exhaustiveness of variant keys given a selector.
+* Type-check values passed into functions.
+* Validate that matching functions are only called in selectors.
+* Validate that formatting functions are only called in placeholders.
+* Forbid edits to certain function options (e.g. currency options).
+* Autocomplete function and option names.
+* Display on-hover tooltips for function signatures with documentation.
+* Display/edit known message metadata.
+* Restrict input in GUI by providing a dropdown with all viable option values.
+
+## Data Model
+
+The registry contains descriptions of function signatures.
+[`registry.dtd`](./registry.dtd) describes its data model.
+
+The main building block of the registry is the `<function>` element.
+It represents an implementation of a custom function available to translation at runtime.
+A function defines a human-readable _description_ of its behavior
+and one or more machine-readable _signatures_ of how to call it.
+Named `<pattern>` elements can optionally define regex validation rules for input, optional values, and variant keys.
+
+The `<signature>` element defines one particular signature of the custom function,
+i.e. the set of arguments and named options that can be used together in a single call to the function.
+The `type` attribute specifies the calling context of the signature.
+A `type="format"` function can only be called inside a placeholder inside translatable text.
+A `type="match"` function can only be called inside a selector.
+Signatures with a non-empty `locales` attribute are locale-specific and only available in translations in the given languages.
+
+A signature may define the type of input it accepts with zero or more `<input>` elments.
+In MessageFormat 2.0 functions can only ever accept at most one positional argument.
+Multiple `<input>` elements can be used to define different validation rules for this single argument input,
+together with appropriate human-readable descriptions.
+
+A signature may also define one or more `<param>` elements representing _named options_ to the function.
```

Comment:
As these are _named options_, why is the element called `<param>` and not e.g. `<option>`?

Reply:
Do we actually refer to the "argument" as such anywhere in the current spec? I've no strong opinions on that, though `option` is clearly preferable over `param`.